File: common/src/main/java/juuxel/adorn/loot/GameRuleLootCondition.java
package juuxel.adorn.loot;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import juuxel.adorn.lib.registry.Registered;
import juuxel.adorn.util.Logging;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.level.gamerules.GameRule;
import net.minecraft.world.level.storage.loot.LootContext;
import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
import org.slf4j.Logger;

public record GameRuleLootCondition(GameRule<?> gameRule) implements LootItemCondition {
    public static final MapCodec<GameRuleLootCondition> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
        BuiltInRegistries.GAME_RULE.byNameCodec()
            .fieldOf("game_rule")
            .forGetter(GameRuleLootCondition::gameRule)
    ).apply(builder, GameRuleLootCondition::new));
    private static final Logger LOGGER = Logging.logger();

    @Override
    public boolean test(LootContext lootContext) {
        var rule = lootContext.getLevel().getGameRules().get(gameRule);

        if (rule instanceof Boolean b) {
            return b;
        } else {
            LOGGER.error("Game rule {} ({}) is not a boolean", rule, gameRule);
        }

        return false;
    }

    @Override
    public MapCodec<? extends LootItemCondition> codec() {
        return CODEC;
    }

    public static Builder builder(GameRule<?> gameRule) {
        return new Builder(gameRule);
    }

    public static Builder builder(Registered<? extends GameRule<?>> gameRule) {
        return builder(gameRule.get());
    }

    public static final class Builder implements LootItemCondition.Builder {
        private final GameRule<?> gameRule;

        private Builder(GameRule<?> gameRule) {
            this.gameRule = gameRule;
        }

        @Override
        public LootItemCondition build() {
            return new GameRuleLootCondition(gameRule);
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/loot/GameRuleLootCondition.kt
package juuxel.adorn.loot

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import juuxel.adorn.util.logger
import net.minecraft.loot.condition.LootCondition
import net.minecraft.loot.condition.LootConditionType
import net.minecraft.loot.context.LootContext
import net.minecraft.world.GameRules
import net.minecraft.world.GameRules.BooleanRule

class GameRuleLootCondition(private val gameRule: String) : LootCondition {
    private val key: GameRules.Key<*> = GameRules.Key(gameRule, GameRules.Category.MISC)

    override fun test(context: LootContext): Boolean =
        when (val rule = context.world.gameRules.get(key)) {
            is BooleanRule -> rule.get()

            null -> {
                LOGGER.error("Unknown game rule {} in loot condition", gameRule)
                false
            }

            else -> {
                LOGGER.error("Game rule {} ({}) is not a boolean", rule, gameRule)
                false
            }
        }

    override fun getType(): LootConditionType =
        AdornLootConditionTypes.GAME_RULE

    companion object {
        val CODEC: Codec<GameRuleLootCondition> = RecordCodecBuilder.create { builder ->
            builder.group(
                Codec.STRING.fieldOf("game_rule").forGetter { it.gameRule }
            ).apply(builder, ::GameRuleLootCondition)
        }
        private val LOGGER = logger()
    }
}
